File: onepass.py
import os
import time

kdominating = []
notkdominating = []
comparisons = 0

def kdominate(obj1, obj2, k):
	global comparisons
	dim = len(obj1)
	comparisons += 1
	greater = False
	kcount = 0


	for index in range(1,dim):
		if obj1[index] < obj2[index] :
			greater = True
		if obj1[index] <= obj2[index]:
			kcount += 1 

	if greater and kcount >= k :
		return True
	else :
		return False

def checkSame(obj1, obj2):
	for index in range(1, len(obj1)):
		if not obj1[index] == obj2[index]:
			return False
	return True

def onepass(infilename, outfilename, k):
	global kdominating
	global notkdominating

	inputfile = open(infilename, 'r')
	for line in inputfile:
		obj = line.rstrip().lstrip().split('\t')
		obj = map(float, obj)

		isUniqueSkyline = True

		for point in notkdominating[:]:
			if kdominate(obj, point, k):
				notkdominating.remove(point)
			elif kdominate(point, obj, k) or checkSame(point, obj) :
				isUniqueSkyline = False
				break

		if isUniqueSkyline:
			isDominant = True
			for point in kdominating[:]:
				if kdominate(point, obj, k):
					isDominant = False
				if kdominate(obj, point, k):
					kdominating.remove(point)
					notkdominating.append(point)

			if isDominant:
				kdominating.append(obj)
			else:
				notkdominating.append(obj)

if __name__ == '__main__':
	startTime = time.time()
	infilename = 'sample_ant.txt'
	outfilename = 'output_1pass.txt'
	k = 5
	onepass(infilename, outfilename,  k)
	
	global kdominating

	kdominatingIdx = []
	for item in kdominating:
		kdominatingIdx.append( int(item[0]))

	endTime = time.time()
	outfile = open(outfilename, 'w')
	outfile.write("Total running time: "+ str(endTime - startTime) + " sec\n")
	outfile.write("Comparisons: "+ str(comparisons)+"\n")
	outfile.write("Size of k-dominating skyline set: "+str(len(kdominatingIdx)) + "\n")
	outfile.write("Ids of the skyline objects: \n")
	outfile.write(str(kdominatingIdx))
	outfile.write("\n")
	outfile.close()

File: onepass.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <sstream> 
#include <fstream> 
using namespace std;
vector< vector<float> > kdominating;
vector< vector<float> > notkdominating;
int kparam, dimension;
int comparisons = 0;
int comparePoints(vector<float> obj1, vector<float> obj2) {
	// comparisons += 1;
	int a=0,b=0,c=0;
	for(int k=1;k<=dimension;k++) {
		if(obj1[k]<obj2[k])
			a++;
		else if(obj2[k]<obj1[k])
			b++;
		else
			c++;
	}
	int res=0;
	if(a && a+c>=kparam)
		res++;
	if(b && b+c>=kparam)
		res+=2;
	return res;
}
bool kdominate(vector<float> obj1, vector<float> obj2){
	bool greater = false;
	int kcount = 0;
	for(int index=1; index<= dimension; index++){
		if(obj1[index] < obj2[index]){
			greater = true;
		}
		if(obj1[index] <= obj2[index]){
			kcount++;
		}
	}

	if(greater && kcount >= kparam){
		return true;
	} else{
		return false;
	}
}

bool dominating(vector<float> obj1, vector<float> obj2){
	bool flag = false;
	for(int index=1; index<= dimension; ++index){
		if(obj1[index] < obj2[index])
			flag = true;
		else if(obj1[index] > obj2[index]){
			return false;
		}
	}
	return flag;
}

bool checkSame(vector<float> obj1, vector<float> obj2){
	for(int index=1; index<=dimension; index++){
		if(obj1[index]!=obj2[index]) return false;
	}
	return true;
}

void onepass(string infile){
	ifstream file(infile.c_str());
	string line;
	while (getline(file, line)){
	    vector<float> obj;
		istringstream iss(line);
	    float a;
	    for(int i=0;i<= dimension;i++){
			iss >> a;
			obj.push_back(a);
		}
		bool isUniqueSkyline = true;
		bool pruning = false;
		vector< vector<float> >::iterator it;
		for(it=notkdominating.begin();it!=notkdominating.end(); ){
			bool flag = true;
			comparisons += 1;
			if(dominating(obj, *it)){
				it = notkdominating.erase(it);
				flag = false;
			} else if(dominating(*it, obj)){
				isUniqueSkyline = false;
				break;
			} 
			if(comparePoints(*it, obj) == 1){
				pruning = true;
			}
			if(flag) ++it;
		}
		if(isUniqueSkyline){
			bool isDominant = true;
			vector< vector<float> >::iterator it;
			for(it=kdominating.begin(); it!=kdominating.end();){
				int c = comparePoints(*it, obj);
				comparisons += 1 ;
				// if(kdominate(*it, obj)){
				// 	isDominant = false;
				// }
				// if(kdominate(obj, *it)){
				// 	notkdominating.push_back(*it);
				// 	it = kdominating.erase(it);
				// } else{
				// 	++it;
				// }
				if(c&1){
					isDominant = false;
				}
				if(c>1){
					notkdominating.push_back(*it);
					it = kdominating.erase(it);
				} else{
					++it;
				}
			}
			if(!pruning && isDominant){
				kdominating.push_back(obj);
 			} else{
 				notkdominating.push_back(obj);
 			}
		}
	}
}

int main(){
	string infilename = "sample_ant.txt";
	string outfilename = "output_1pass.txt";
	clock_t t1,t2;
    t1=clock();
	kdominating.clear();
	notkdominating.clear();
	kparam = 5;
	dimension = 5;
	onepass(infilename);
	t2=clock();
    float diff ((float)t2-(float)t1);

	vector< vector<float> >::iterator it;
	vector<int> result;
	for(it=kdominating.begin(); it!=kdominating.end(); ++it){
		std::vector<float> obj = *it;
		result.push_back(int (obj[0])); 
	}
	sort(result.begin(), result.end());
	ofstream myfile;
	myfile.open(outfilename.c_str());
	myfile << "Time taken : " << ((float)diff)/CLOCKS_PER_SEC << " seconds" <<endl;
	myfile << "Comparisons : " << comparisons << endl;
	myfile << "Size of k-dominating skyline set: " << result.size() << endl;

	vector<int>::iterator iter;
	for(iter=result.begin(); iter!=result.end(); ++iter){
		myfile << *iter << ", ";
	}
	myfile << endl;
	myfile.close();
	return 0;
}